### PR TITLE
Allow external grader result processing to be stopped

### PR DESCRIPTION
**Pull request by nwalters512**
This will be used in #6381; I'm submitting it via a separate PR to make reviewing easier.

### CODE REVIEW

File: lib/externalGraderResults.js [hunk 1]
```diff
@@ -11,129 +10,116 @@ const externalGradingSocket = require('./externalGradingSocket');
...
+  if (Number.isNaN(jobId)) {
+    throw error.makeWithData('Message does not contain a valid grading job id.', data);
   }
 
-  var logBundle = [];
-  var loggingTimeout = setTimeout(logAtTimeout, 10000, jobId, logBundle);
+  const logBundle = [];
+  const loggingTimeout = setTimeout(() => {
+    logger.info('externalGradingResults timeout', { jobId, logBundle });
+  }, 10000);
```
Comment (nwalters512) [inline lib/externalGraderResults.js:122]: What exactly is this log bundle stuff? I can see it was introduced in #1868, but that doesn't include any description of the problem we're trying to solve. Were we experiencing results that took a long time to process? Do you think this is still necessary?

File: lib/externalGraderResults.js [hunk 2]
```diff
@@ -11,129 +10,116 @@ const externalGradingSocket = require('./externalGradingSocket');
...
-          },
-        ],
-        (err) => {
-          if (ERR(err, (err) => logger.error('Error handling grader results', err)));
-          next();
+  const queueUrl = await loadQueueUrl(sqs);
+
+  // Start work in an IIFE so we can keep going asynchronously
+  // after we return to the caller.
+  (async () => {
```
Comment (nwalters512) [inline lib/externalGraderResults.js:28]: I ended up rewriting this file in `async`/`await`. The core thing we needed was the ability to actually wait for processing to finish; previously we would get results from SQS and then essentially fire-and-forget the actual processing. Now that we specifically need to wait for processing to stop, we need to propagate the notion of "done processing" back up. I could have done that wil callbacks, but it was only marginally more work to rewrite everything 🤷 

We'll have to make sure we test this carefully.

**Review by mwest1066 — APPROVED** [review]

File: lib/externalGrader.js [hunk 3]
```diff
@@ -98,8 +100,10 @@ module.exports._beginGradingJob = function (grading_job, submission, variant, qu
   gradeRequest.on('results', (gradingResult) => {
     // This event will only be fired when running locally; in production,
     // external grader results wil be delivered via SQS.
-    assessment.processGradingResult(gradingResult);
-    logger.verbose(`Successfully processed grading job ${grading_job.id}`);
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply lib/externalGrader.js:102]: Exactly: just a bug.

File: lib/externalGrader.js [hunk 4]
```diff
@@ -70,7 +70,9 @@ module.exports._beginGradingJob = function (grading_job, submission, variant, qu
     };
 
     // Send the grade out for processing and display
-    assessment.processGradingResult(ret);
+    assessment
+      .processGradingResult(ret)
+      .catch((err) => logger.error('Error processing grading result', err));
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply lib/externalGrader.js]: Done in f6227ee.

File: lib/externalGrader.js [hunk 5]
```diff
@@ -121,7 +125,9 @@ function handleGraderError(jobId, err) {
       },
     },
   };
-  assessment.processGradingResult(gradingResult);
+  assessment
+    .processGradingResult(gradingResult)
+    .catch((err) => logger.error('Error processing grading result', err));
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply lib/externalGrader.js]: Done in f6227ee.

File: lib/externalGraderResults.js (hunk 1 above)
Comment (nwalters512) [reply lib/externalGraderResults.js]: Done in 302bacd.

File: lib/externalGrader.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwest1066) [reply lib/externalGrader.js:102]: Cool, thanks for confirming.

**Pull request merged**